### PR TITLE
Utilize size and capacity for optimizing string allocation and copying

### PR DESCRIPTION
**Pull request by assem2002**
Towards #2257 
- Introduce extend_string() --> To only extend strings with correct capacity that can hold the required size.
- Utilize lfortran_strcpy_descriptor_string --> to either allocate or extend the string.
- Utilize lfortran_strcpy_pointer_strign --> to just copy from src to dest
- Utilize lfortran_alloc --> To create initial capacity for descriptor strings(allocatables)
- Utilize lfortran_string_write --> To copy string appropriatlely

### CODE REVIEW

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 1]
```diff
@@ -2008,29 +2008,91 @@ LFORTRAN_API void _lfortran_strcat(char** s1, char** s2, char** dest)
...
 }
+// Allocate + Extend String ----------------------------------------------------------- 
+
+void extend_string(char** ptr, int32_t new_size /*Null-Character Included*/,
+    int64_t* current_size, int64_t* string_capacity){
+    ASSERT_MSG(current_size != NULL,"%s", "string size is NULL");
+    ASSERT_MSG(string_capacity != NULL, "%s", "string capacity is NULL");
+    ASSERT_MSG(*string_capacity >= *current_size, "%s", "compiler-behavior error : string capacity < string size");
+/* Remove this (It's for debugging purposes) */
+    ASSERT_MSG(strlen(*ptr) == (*current_size),"%s","Compiler-Internal Error : Size in string descriptor doesn't match actual size of char*");
```
Comment (assem2002) [inline src/libasr/runtime/lfortran_intrinsics.c:2019]: I think this assertion should be removed as `strlen` would be expensive for long strings. We even could utilize the size in the descriptor string to be the return for such statement
```fortran
character(:), allocatable :: str
print *, len(str) ! HERE
```

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 2]
```diff
@@ -2008,29 +2008,91 @@ LFORTRAN_API void _lfortran_strcat(char** s1, char** s2, char** dest)
...
+            }
+            *ptr = (char*)malloc(inital_capacity);
+            *string_capacity = inital_capacity;
+            const int8_t null_terminated_char_len = 1;
+
+            *string_size = desired_size - null_terminated_char_len;
+        } else if(*ptr != NULL && *string_capacity != 0){
+            // printf("runtime error: Attempting to allocate already allocated variable\n");
+            // exit(1);
+        } else {
```
Comment (assem2002) [inline src/libasr/runtime/lfortran_intrinsics.c:2052]: This should be uncommented, but we face an issue with correct deallocations. I'll open an issue with an MRE. Once we can fix the MRE, We can uncomment this line.

File: src/libasr/runtime/lfortran_intrinsics.c [hunk 3]
```diff
@@ -2008,29 +2008,91 @@ LFORTRAN_API void _lfortran_strcat(char** s1, char** s2, char** dest)
...
-            x[0][i] = ' ';
+            extend_string(x, y_len+1, x_string_size, x__string_capacity);
+        }
+    }
+
+    if(!is_descriptor_string){
+        x_len = strlen(*x); // Stick to LHS-string-fixed size.
+    } else {
+        x_len = y_len; // LHS follows RHS size. 
+    }
```
Comment (certik) [inline src/libasr/runtime/lfortran_intrinsics.c:2084]: I don't like this logic if it is descriptor or not.

In ASR, the assignment should only be between two fixed size strings, or two descriptor strings.

So in the backend, just write two functions: "copy_fixed_size" and "copy_descriptor", and in the LLVM backend dispatch correctly.

Comment (assem2002) [reply src/libasr/runtime/lfortran_intrinsics.c:2084]: I like that approach.

File: src/libasr/runtime/lfortran_intrinsics.c (hunk 2 above)
Comment (assem2002) [reply src/libasr/runtime/lfortran_intrinsics.c:2052]: Related to this #5249. Once that is fixed we can turn that on.

File: src/libasr/runtime/lfortran_intrinsics.c (hunk 3 above)
Comment (certik) [reply src/libasr/runtime/lfortran_intrinsics.c:2084]: I think this looks a lot better now!

**Comment by certik** [comment]
I think it looks in the right direction now.

**Review by gxyd — APPROVED** [review]
This looks good to me.

@assem2002 , are you going to push anymore changes to this?

**Review by certik — APPROVED** [review]

**Pull request merged**